test/test_HeadingController.py

```diff
@@ -0,0 +1,354 @@
+#!/usr/bin/env python
+import local_imports
+import rostest
+import unittest
+import sailbot_constants
+from heading_controller import HeadingController
+from control_modes import ControlModes
+
+# Do something with local_imports to avoid lint errors
+local_imports.printMessage()
+
+
+class Test_HeadingController(unittest.TestCase):
+
+    # Test for bad ControlMode input
+    def test_badControlModeInput_ValueError(self):
+        with self.assertRaises(ValueError):
+            HeadingController(0.1, 0, -1)
+
+    # Test for switching on the switch interval
+
+    def test_switch_interval(self):
+        mock_speed = sailbot_constants.SPEED_THRESHOLD_FOR_JIBING_KNOTS + 0.1
+        mock_time = 0
+        mock_heading_error = 2 * sailbot_constants.MIN_HEADING_ERROR_FOR_SWITCH
+
+        # Enter TACKABLE mode from UNKNOWN
+        hc = HeadingController(mock_speed, mock_time, ControlModes.UNKNOWN.value)
+
+        # No switch
+        self.assertFalse(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+        # Adjust parameters
+        mock_time = 0.5 * sailbot_constants.SWITCH_INTERVAL
+        mock_heading_error = 0.5 * sailbot_constants.MIN_HEADING_ERROR_FOR_SWITCH
+
+        # Even if exit condition is met for TACKABLE, no switch until switch interval timeout
+        self.assertFalse(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+        # Adjust parameters
+        mock_time = sailbot_constants.SWITCH_INTERVAL
+
+        # Switch occurs
+        self.assertTrue(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+        # Adjust time
+        mock_time = 1.5 * sailbot_constants.SWITCH_INTERVAL
+
+        # Even if exit condition is met for TACKABLE, no switch until switch interval timeout
+        self.assertFalse(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+        # Adjust parameters
+        mock_time = 2 * sailbot_constants.SWITCH_INTERVAL
+
+        # Switch occurs
+        self.assertTrue(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+    # TACKABLE tests
+
+    # Entry to TACKABLE from UNKNOWN
```

Comment:
don't think we need this comment (or we can rename the testMethod to capture this exact scenerio)? It's actually totally ok to have really long test method names (as long as it's below the 120 character line limit)